### ISSUE TITLE
Adjusting and saving label text size

### ISSUE THREAD

**Issue by ericsturman** [TeselaGen/openVectorEditor#648]
@tnrich

We have been receiving quite a few requests for a way to adjust the text size of labels in the circular view.  Users would like to scale the relative size of text to avoid the labels that show the number of additional labels not shown (ex: `+3,Eco065I` as shown in the picture).  I looked into making the change a while back, but it wasn't immediately obvious how to do it.  Ultimately I think that if users can adjust the scale of the text and save the setting, it will work fine for most of the vectors they look at.  I'm not sure of the magnitude that they would want to change it, so making it pretty flexible would be great.  

Any thoughts?

Thanks,
Eric

![Screen Shot 2020-09-16 at 1 38 43 PM](https://user-images.githubusercontent.com/4339551/93378654-5c56f280-f822-11ea-8d89-483ad3435151.png)

**Comment by tnrich** [comment]
Hey @ericsturman that does sound like a good feature! I am a little busy at the moment but can probably look into it sometime in the future. If you want to take a crack at doing it I would be happy to merge in a PR. 

Just thinking about it aloud.. I would imagine adding a "Adjust Circular View Label Size" menu to the View Menu with "Increase Label Size" and "Decrease Label Size" sub options. The user-set label size value should be stored in redux and synced with localstorage. 

You'll need to play around with the actual label height/width logic. I think most of it is abstracted out into constants that will need to instead be passed as props. I believe relaxLabelAngels.js holds the logic for stopping the labels from running into eachother.  

I can take a deeper look into what would need to happen logic-wise when I get a moment. 

What do you think about that plan generally?

-Thomas

**Comment by ericsturman** [comment]
This sounds like a good plan @tnrich .  I will be happy to take a look at the code again and see what I can accomplish and let you know.
Eric

**Comment by tnrich** [comment]
@ericsturman any progress on this one? :)

**Comment by ericsturman** [comment]
Hi @tnrich .  I'm able to adjust the text size, but I'm having difficulty figuring out how labels get clumped together in each quadrant.  It seems like the upper right quadrant should be split up so that the label line for `part 0` doesn't intersect with the part.

Eric

![label_line_and_grouping](https://user-images.githubusercontent.com/4339551/93948872-2bd0f600-fd05-11ea-9cc3-b074b2b24b1c.png)

**Comment by tnrich** [comment]
@ericsturman can you push a branch for me to play around on?

**Comment by ericsturman** [comment]
@tnrich I submitted a pull request... maybe you only wanted a branch name on my forked repo?

**Comment by tnrich** [comment]
Oh, that's great I'll take a look!

**Comment by ericsturman** [comment]
@tnrich I'm finally seeing that the issue is the following `if` statement in `repositionAndGroupLabels` function in `relaxLabelAngles` ... I think adjusting the logic for calculating `naturalSlot` will help:
```
          if (naturalSlot > extraSpaces) {
            label.y = lastLabelYPosition;
          }
```

**Comment by tnrich** [comment]
@ericsturman yeah, I think you should play with the label positioning logic. I believe the general outline of how it works is the following: 

- split labels into the 4 quadrants (top left, top right, bottom left, bottom right) 
- for each quadrant, start at the bottom (or top if it is a bottom quadrant) and place each label above the previous one until all the labels in that quadrant have been placed
- if there are still labels that need to be place but we've run out of height, group them together in a label group 

I think the way to change that to get a nicer spacing would be the following: 
Instead of placing each label above the previous one, first calculate the number of slots (slots = quadrant height / label height) 
Then if are _fewer_ labels than slots for a quadrant, we should increase the slot spacing so that the labels are as spaced out as possible. 
If there are _more_ labels than slots for a quadrant we shouldn't do anything :) 

Does that plan of attack make sense to you? I think it would address the issue you're seeing with the part label being positioned strangely (just by having a better spacing strategy in general).

**Comment by ericsturman** [comment]
@tnrich Yes...that's how I understand the positioning logic to work... and I agree that's a good approach.

**Comment by tnrich** [comment]
@ericsturman awesome I'll check it out!

**Comment by tnrich** [comment]
Looks good to me! If you could just write a test for it asserting that when the label size preference is changed that the label font size changes that would great. I'll merge and publish after that 👍

**Comment by ericsturman** [comment]
@tnrich I have added tests for label size, label spacing and label line intensity (from an earlier pull request).

**Comment by tnrich** [comment]
@ericsturman I published open-vector-editor@10.2.4 that should include your changes. Nice work!
